### DISCUSSION TITLE
Image reflector generates 401 requests to /v2 endpoint

### DISCUSSION THREAD

**Discussion by mecseid** [fluxcd/flux2#2893] [Q&A] [answered]
Hi,

The image-reflector-controller generates me a "lot of" request to /v2 API endpoint without any authorization header, and it will generates 401 responses.

I don't know how to debug this kind of issue,, so because of that ask you for help. :(

Generated log message by docker registry:
```
{"@timestamp":"2022-07-04T20:41:50.626691862Z","@version":"1","go.version":"go1.16.15","http.request.host":"docker.test.local","http.request.id":"05bf558c-4f8c-48af-8879-ea7747d9bb19","http.request.method":"GET","http.request.remoteaddr":"10.142.0.33","http.request.uri":"/v2/","http.request.useragent":"Go-http-client/2.0","level":"warning","message":"error authorizing context: basic authentication challenge for realm \"test\": invalid authorization credential"}
10.142.162.191 - - [04/Jul/2022:20:41:50 +0000] "GET /v2/ HTTP/1.1" 401 87 "" "Go-http-client/2.0"
```

**Comment by kingdonb** [comment]
Are these errors in your image registry or in the image-reflector-controller logs?

The /v2 API endpoint is not a valid endpoint that you can access in the docker registry API as far as we can tell. If you have loaded some ImageRepository manifests onto your cluster in order to arrive at this state, can you please share them?

Thanks for reporting the trouble.

**Reply by mecseid** [reply]
The errors came from the image registry (https://github.com/distribution/distribution).
If I scale down the `image-reflector-controller`, then these type of API calls dissapear.

At tbe moment I have 30 `ImageRepository` maniftest, and all condition reason is `ReconciliationSucceeded`.
```shell
$ kubectl get -A imagerepositories.image.toolkit.fluxcd.io -o json | jq '.items[].status.conditions[].reason' | sort | uniq -c
     30 "ReconciliationSucceeded"
```

I don't know how to continue the debugging, but I will send anything if necessary.
Maybe can I produce some kind of debug log in image-reflector-controller about API calls (request/response)?

**Reply by mecseid** [reply]
Hey @kingdonb !
A little update. I got the following Slack message from notification controller:
```
imagerepository/onlineinvoice.infra-cicd-gitops
Get "https://nfdocker.local/v2/": http2: client connection force closed via ClientConn.Close
```

The belonging CRD:
```yaml
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageRepository
metadata:
  creationTimestamp: "2022-04-26T21:23:49Z"
  finalizers:
  - finalizers.fluxcd.io
  generation: 5
  labels:
    kustomize.toolkit.fluxcd.io/name: application-ops
    kustomize.toolkit.fluxcd.io/namespace: infra-cicd-gitops
  name: onlineinvoice
  namespace: infra-cicd-gitops
  resourceVersion: "620922715"
  uid: 40eda41a-7dbd-4052-95c3-373252d6c0c0
spec:
  image: nfdocker.local/onlineinvoice
  interval: 1m0s
  secretRef:
    name: nexus-nfdocker
status:
  canonicalImageName: nfdocker.local/onlineinvoice
  conditions:
  - lastTransitionTime: "2022-07-09T20:53:54Z"
    message: successful scan, found 2 tags
    reason: ReconciliationSucceeded
    status: "True"
    type: Ready
  lastScanResult:
    scanTime: "2022-07-11T16:15:30Z"
    tagCount: 2
  observedGeneration: 5
```

The used image-reflector: `ghcr.io/fluxcd/image-reflector-controller:v0.19.2`

**Reply by kingdonb** [reply]
I am not sure what causes the message:

```
error authorizing context: basic authentication challenge for realm \"test\": invalid authorization credential"
```
in the Docker registry, but I would look for any ImageRepository resources on the cluster that are improperly created with the word "test" in the image. If someone created this ImageRepository with no authorization secret and the "test" realm is something that does not really exist in a public form on the harbor instance, or if the "test" imagerepository is created with an invalid authorization secret for the "test" realm", then it seems likely that it would produce error messages like this one.

~If you delete the failing test ImageRepository then I think you'll see those authorization failure messages go away.~

**Reply by kingdonb** [reply]
I have been using image reflector and friends, and I see the error `"GET /v2/ HTTP/2.0" 401 76 "-"` in my nginx logs now, it repeats every 60s and on my low-traffic cluster which does not have any other active clients on the Harbor registry besides my one cluster, it makes up roughly 67% of the traffic that hits my Harbor instance (the only persistent error):

![Screen Shot 2022-07-21 at 10 40 52 AM](https://user-images.githubusercontent.com/3286998/180242129-02718091-9ebd-48bf-9ea0-a7417f2770d3.png)

I see why this is annoying now.

**Reply by kingdonb** [reply]
I thought it might have something to do with the authentication status of the `ImageRepository` – whether it has a secret attached or not – and it seems it does not. This is the conversation from the Harbor side, or from the Ingress-nginx side in between at least, every time an ImageRepository is reconciled:

```
127.0.0.1 - - [21/Jul/2022:15:06:02 +0000] "GET /v2/ HTTP/2.0" 401 76 "-" "Go-http-client/2.0" 24 0.001 [vcluster-moo-cluster-mgorr-harbor-core-x-harbor-x-moo-cluster-80] [] 10.36.0.39:8080 76 0.000 401 e32765032886ec96b5ef01151fffb513
127.0.0.1 - robot$planetstore+planet-readonly [21/Jul/2022:15:06:02 +0000] "GET /service/token?scope=repository%3Aplanetstore%2Fweb%3Apull&service=harbor-registry HTTP/2.0" 200 978 "-" "go-containerregistry/v0.9.0" 7 0.012 [vcluster-moo-cluster-mgorr-harbor-core-x-harbor-x-moo-cluster-80] [] 10.36.0.39:8080 978 0.012 200 a5bdd20cd8e1eeb5ad49667d52c30a4d
127.0.0.1 - - [21/Jul/2022:15:06:02 +0000] "GET /v2/planetstore/web/tags/list?n=1000 HTTP/2.0" 200 1575 "-" "go-containerregistry/v0.9.0" 924 0.002 [vcluster-moo-cluster-mgorr-harbor-core-x-harbor-x-moo-cluster-80] [] 10.36.0.39:8080 1575 0.000 200 3a7ed5835f4ec5bcb2458787eaf9f14f

127.0.0.1 - - [21/Jul/2022:15:06:04 +0000] "GET /v2/ HTTP/2.0" 401 76 "-" "Go-http-client/2.0" 17 0.001 [vcluster-moo-cluster-mgorr-harbor-core-x-harbor-x-moo-cluster-80] [] 10.36.0.39:8080 76 0.000 401 f15bbd40451510f1b7ff66ed0ac11c5a
127.0.0.1 - - [21/Jul/2022:15:06:04 +0000] "GET /service/token?scope=repository%3Aexamples%2Fsinatra-mustache%3Apull&service=harbor-registry HTTP/2.0" 200 946 "-" "go-containerregistry/v0.9.0" 73 0.009 [vcluster-moo-cluster-mgorr-harbor-core-x-harbor-x-moo-cluster-80] [] 10.36.0.39:8080 946 0.012 200 f4f7288bfd92042b241d8e3fd365dc76
127.0.0.1 - - [21/Jul/2022:15:06:04 +0000] "GET /v2/examples/sinatra-mustache/tags/list?n=1000 HTTP/2.0" 200 1384 "-" "go-containerregistry/v0.9.0" 926 0.002 [vcluster-moo-cluster-mgorr-harbor-core-x-harbor-x-moo-cluster-80] [] 10.36.0.39:8080 1384 0.004 200 dc1de88ee399b8f18125c10c8d438d53
```

**Comment by kingdonb** [comment, marked as answer]
I'm not sure how `1x 401` + `2x 200` responses works out to 66% failure rate but that's what the chart says, and I'd love it if the normal operation didn't trigger any error responses at all. Is this a Harbor bug or Flux? (Is there any reason to call `/v2/` at all? I couldn't find the relevant code as I do not understand the total function of the image reflector controller myself)

**Reply by mecseid** [reply]
I don't think it is a Harbor related bug, because I got these failure responses with Docker Registry too. 😞

**Reply by kingdonb** [reply]
I've done some experimenting with the raw messages that are passed back and forth when a registry connection is made (in a pure ruby library) and found that a 401 response from `/v2/` is a normal part of the conversation, it contains the information about how to request a service token (bearer token) and what realm to use.

This cannot be fixed, it's apparently a critical part of the conversation and service token authentication cannot complete without it.
